domain/modelconfig/service/service.go

```diff
@@ -295,19 +298,22 @@ func (s *Service) UpdateModelConfig(
 // secrets in dqlite
 type dummySecretsBackendProvider struct{}
 
-// dummySpaceProvider implements validators.SpaceProvider and always returns true.
-// TODO (tlm): These needs to be swapped out with an actual checker when we have
-// spaces in dqlite
-type dummySpaceProvider struct{}
-
 // HasSecretsBackend implements validators.SecretBackendProvider
 func (*dummySecretsBackendProvider) HasSecretsBackend(_ string) (bool, error) {
 	return true, nil
 }
 
-// HasSpace implements validators.SpaceProvider
-func (*dummySpaceProvider) HasSpace(_ string) (bool, error) {
-	return true, nil
+// spaceValidator implements validators.SpaceProvider.
+type spaceValidator struct {
+	st State
+}
+
+// HasSpace implements validators.SpaceProvider. It checks whether the
+// given space exists.
+func (v *spaceValidator) HasSpace(spaceName string) (bool, error) {
+	// TODO(nvinuesa): We should inject the context into this function,
+	// which will come from the environs ValidateFunc.
+	return v.st.CheckSpace(context.TODO(), spaceName)
```

Comment:
I think we should have the context passed in on construction. I would also consider having a function that returns a closure implementation so the context feels like it is being passed to func and not held by a long running struct.

Reply:
I answered in https://github.com/juju/juju/pull/17330#issuecomment-2092969819